GDJS/Runtime/websocket-debugger-client/hot-reloader.js

```diff
@@ -0,0 +1,1313 @@
+    } else {
+      if (oldBehaviorConstructor !== newBehaviorConstructor) {
+        this._logs.push({
+          kind: 'info',
+          message:
+            'Behavior with type ' +
+            behaviorTypeName +
+            ' was changed, and will be re-instantiated in gdjs.RuntimeObjects using it.',
+        });
+        changedRuntimeBehaviors.push({
+          oldBehaviorConstructor,
+          newBehaviorConstructor,
+          behaviorTypeName,
+        });
+      }
+    }
+  }
+  return changedRuntimeBehaviors;
+};
+
+/**
+ * @param {ProjectData} newProjectData
+ * @param {RuntimeGameOptionsScriptFile[]} oldScriptFiles
+ * @param {RuntimeGameOptionsScriptFile[]} newScriptFiles
+ * @param {boolean} projectDataOnlyExport
+ */
+gdjs.HotReloader.prototype.reloadScriptFiles = function (
+  newProjectData,
+  oldScriptFiles,
+  newScriptFiles,
+  projectDataOnlyExport
+) {
+  const reloadPromises = [];
+
+  // Reload events, only if they were exported.
+  if (!projectDataOnlyExport) {
+    newProjectData.layouts.forEach((_layoutData, index) => {
+      reloadPromises.push(this._reloadScript('code' + index + '.js'));
+    });
+  }
+
+  for (let i = 0; i < newScriptFiles.length; ++i) {
+    const newScriptFile = newScriptFiles[i];
+    const oldScriptFile = oldScriptFiles.filter(
+      (scriptFile) => scriptFile.path === newScriptFile.path
+    )[0];
+    if (!oldScriptFile) {
+      // Script file added
+      this._logs.push({
+        kind: 'info',
+        message:
+          'Loading ' +
+          newScriptFile.path +
+          ' as it was added to the list of scripts.',
+      });
+      reloadPromises.push(this._reloadScript(newScriptFile.path));
+    } else {
+      // Script file changed, which can be the case for extensions created
+      // from the editor, containing free functions or behaviors.
+      if (newScriptFile.hash !== oldScriptFile.hash) {
+        this._logs.push({
+          kind: 'info',
+          message:
+            'Reloading ' + newScriptFile.path + ' because it was changed.',
+        });
+        reloadPromises.push(this._reloadScript(newScriptFile.path));
+      }
+    }
+  }
+  for (let i = 0; i < oldScriptFiles.length; ++i) {
+    const oldScriptFile = oldScriptFiles[i];
+    const newScriptFile = newScriptFiles.filter(
+      (scriptFile) => scriptFile.path === oldScriptFile.path
+    )[0];
+    // A file may be removed because of a partial preview.
+    if (!newScriptFile && !projectDataOnlyExport) {
+      this._logs.push({
+        kind: 'warning',
+        message: 'Script file ' + oldScriptFile.path + ' was removed.',
+      });
+    }
+  }
+
+  // @ts-ignore - Promise not supported in ES5 mode.
+  return Promise.all(reloadPromises);
+};
+
+/**
+ * @param {ProjectData} oldProjectData
+ * @param {ProjectData} newProjectData
+ * @param {ChangedRuntimeBehavior[]} changedRuntimeBehaviors
+ * @param {gdjs.RuntimeGame} runtimeGame
+ */
+gdjs.HotReloader.prototype._hotReloadRuntimeGame = function (
```

Comment:
@Bouh Look at this file and all the _hotReloadXXX methods that are checking the changes between the old data and the new data, and then calling the game/scene/object/behavior accordingly to reflect the changes.